src/System.Text.Primitives/System/Text/Encoding/Utf16/Utf16TextEncodingLE.cs

```diff
@@ -7,113 +7,145 @@ namespace System.Text.Utf16
-                            && lowBytes >= UnicodeConstants.Utf16LowSurrogateFirstCodePoint
-                            && lowBytes <= UnicodeConstants.Utf16LowSurrogateLastCodePoint)
-                    {
-                        answer = (((highBytes - UnicodeConstants.Utf16HighSurrogateFirstCodePoint) << 10)
-                            | (lowBytes - UnicodeConstants.Utf16LowSurrogateFirstCodePoint)) + 0x10000;
-
-                        decodedCodePoint = new UnicodeCodePoint(answer);
-                        bytesWrittenForCodePoint = 4;
-                    }
-                }
+        public override bool TryEncode(ReadOnlySpan<byte> utf8, Span<byte> data, out int bytesConsumed, out int bytesWritten)
+        {
+            int charactersWritten;
+            var utf16 = data.Cast<byte, char>();
+            var result = Utf8Encoder.TryDecode(utf8, utf16, out bytesConsumed, out charactersWritten);
+
+            bytesWritten = charactersWritten * sizeof(char);
+            return result;
+        }
+
+        public override bool TryEncode(ReadOnlySpan<char> utf16, Span<byte> data, out int charactersConsumed, out int bytesWritten)
+        {
+            // TODO: Other methods validate that the input stream contains valid sequences as they are consumed.
+            //       Currently, this is a copy operation with no validation. What is the right thing here?
 
-                decoded[i] = decodedCodePoint;
-                bytesWritten += bytesWrittenForCodePoint;
+            var charBuffer = data.Cast<byte, char>();
+
+            if (utf16.Length > charBuffer.Length)
+            {
+                charactersConsumed = 0;
+                bytesWritten = 0;
+                return false;
             }
 
+            utf16.CopyTo(charBuffer);
+            bytesWritten = utf16.Length * sizeof(char);
+            charactersConsumed = utf16.Length;
             return true;
         }
 
-        public override bool TryEncodeChar(char value, Span<byte> buffer, out int bytesWritten)
+        public override bool TryEncode(ReadOnlySpan<uint> utf32, Span<byte> data, out int charactersConsumed, out int bytesWritten)
+        {
+            int written;
+            var utf16 = data.Cast<byte, char>();
+            var result = Utf16LittleEndianEncoder.TryEncode(utf32, utf16, out charactersConsumed, out written);
+
+            bytesWritten = written * sizeof(char);
+            return result;
+        }
+
+        public override bool TryEncode(string text, Span<byte> data, out int bytesWritten)
         {
-            if (buffer.Length < 2)
+            bytesWritten = text.Length << 1;
+            if (bytesWritten > data.Length)
             {
                 bytesWritten = 0;
                 return false;
             }
-            buffer[0] = (byte)value;
-            buffer[1] = (byte)(value >> 8);
-            bytesWritten = 2;
+
+            unsafe
```

Comment:
I think we have a Slice extension method for string that would let you do this without unsafe code.